keps/sig-cloud-provider/20200928-extend-konnectivity-for-both-directions.md

```diff
@@ -0,0 +1,194 @@
+
+### Handling the Traffic from the Pods to the Agent
+
+As mentioned above, pods make use of  the Kubernetes default service to reach the KAS. To keep things transparent from a Pod perspective, they will hit the Konnectivity Agent using the Kubernetes default service. The endpoint will be the Konnectivity Agent instead of the KAS. 
+The configuration part of the Kubernetes default service will be done using the Apiserver flag `--advertise-address ip` on the Control Plane side.
+`--advertise-address ip` should match the `--bind-address ip` of the Konnectivity Agent described above. 
+
+### Handling the Traffic from the Kubelet to the Agent
+
+Kubelet does not use the Kubernetes default service to reach the KAS. Instead it relies on a bootstrap kubeconfig file that is used to connect to the KAS. It then generates a proper kubeconfig file that will be using the same URL.
+Instead of specifying the KAS FQDN/address in the bootstrap kubeconfig file, we will be using the local IP address of the Konnectivity agent (`--bind-address ip`).
+
+### Deployment Model
+
+The agent can be run as static pod or systemd units. In any case the agent should be started to give access to the KAS to the kubelet first and to the hosted pods later. This means that using DaemonSets or Deployments is not an option in this setup, because the kubelet would not be able to get the pod manifests from the KAS.
+
+### Listening Interface for Konnectivity agent
+
+As mentioned before, we will be using the Kubernetes default service to route traffic to the Agent. The service in itself has a couple of limitations: it can’t be used as a type externalName, thus preventing the use of dynamic IPs (say, Pods’ networking range). But also, some general services limitations apply: endpoints can’t use the link-local range and the localhost range. This means that we are left with the 3 private IPs ranges (10.0.0.0/8, 172.16.0.0/12 and 192.168.0.0/16). 
+
+When using a StaticPod, the Agent Pod will create a dummy interface (using an init container?), bind to the IP given in the argument `--bind-address` and will start to listen on this IP:local_port (local_port is defined with the `--target` argument).
+With systemd the principle is the same, but the creation and configuration of the interface could be handled with `ExecStartPre` commands.
+
+### Authentication
+
+Konnectivity agent currently support mTLS or Token based authentication. Note that API objects such as Secrets cannot be accessed either when a StaticPod or Systemd service deployment strategy is used. The authentication secret should be made available to the agent through a different channel (e.g. provisioned in the worker node file-system).
```

Comment:
This is talking about the authentication between the proxy agent and the proxy server, right? Can you point this out in the KEP?

Reply:
Yes you are right, we'll make it clear.